Documentation/teaching/labs/networking.rst

```diff
@@ -0,0 +1,1251 @@
+running the :command:`make build` command. The test script is copied to the
+virtual machine only if it is marked as executable. The script uses the
+statically compiled :command:`netcat` tool in :file:`skels/networking/netcat`;
+this executable must have execution permissions.
+
+After running the checker the output should be similar to the one bellow:
+
+.. code-block:: console
+
+  # ./test-2.sh
+  [  797.673535] TCP connection initiated from 127.0.0.1:44721
+  Should show up in filter.
+  Should NOT show up in filter.
+  Check dmesg output.
+
+The test ask for packet filtering first for the ``127.0.0.1`` IP address and
+then for the ``127.0.0.2`` IP address. The first connection initiation packet
+(to ``127.0.0.1``) is intercepted and displayed by the filter, while the second
+(to ``127.0.0.2``) is not intercepted.
+
+3. Listening on a TCP socket
+----------------------------
+
+Write a kernel module that creates a TCP socket that listens to connections on
+port ``60000`` on the loopback interface (in ``init_module``). Start from the
+code in :file:`3-4-tcp-sock` fill in the areas marked with ``TODO 1`` taking
+into account the observations below.
+
+Read the `Operations on the socket structure`_ and `The struct proto_ops
+structure`_ sections.
+
+The ``sock`` socket is a ``server socket`` and must be put in the listening
+state. That is, the ``bind`` and ``listen`` operations must be applied to the
+socket. For the ``bind`` and ``listen`` equivalent, in kernel space you will
+need to call ``sock->ops->...;`` examples of such functions you can call are
+``sock->ops->bind``, ``sock->ops-->listen`` etc.
+
+.. note::
+
+  For example, call ``sock->ops->bind``, or ``sock->ops->listen`` functions, see
+  how they are called in the :c:func:`sys_bind` and :c:func:`sys_listen` system
+  call handlers.
+
+.. note::
+
+  For the second argument of the ``listen`` (backlog) call, use the
+  ``LISTEN_BACKLOG``.
+
+Remember to release the socket in the module's exit function and in the area
+marked with error labels; use :c:func:`sock_release`.
+
+For testing, run the :command:`3-4-tcp_sock/test-3.sh` script. The script is
+copied on the virtual machine by :command:`make copy` only if it is marked as
+executable.
+
+After running the test, a TCP socket will be displayed by listening to
+connections on port ``60000``.
+
+4. Accepting connections in kernel-space
+----------------------------------------
+
+Expand the module from the previous exercise to allow an external connection (no
+message, only new connection accepted). Fill in the areas marked with ``TODO
+2``.
+
+Read the `Operations on the socket structure`_ and `The struct proto_ops
+structure`_ sections.
+
+For the kernel-space ``accept`` equivalent, see the system call handler for
+:c:func:`sys_accept4`. Follow the :c:func:`lnet_sock_accept` implementation, and
+how the ``sock->ops->accept`` call is used. Use ``0`` as the value for the last
+argument (``flags``).
```

Comment:
The last argument is now a bool kern in 4.19.
https://elixir.bootlin.com/linux/v4.19/source/include/linux/net.h#L133

Reply:
You are right. It was an error in translation.